### DISCUSSION TITLE
How can i use tailwind css and mantine together ??

### DISCUSSION THREAD

**Discussion by leamandeep** [mantinedev/mantine#1672] [Q&A] [answered]
Hello Team 👋 !! while working on big project . I'm facing big trouble for every single thing such as notification , modal etc . For All those stuff i have to write it from scratch which is very painful for developer who is handling frontend and backend both . Then i found fireship video on  Mantine (he brief described about mantine) . 

>How can i use Tailwind CSS  and Mantine together in single project??

**Comment by rtivital** [comment, marked as answer]
You can use mantine with tw without any additional setup, just install both libraries. Make sure not to include tw base styles as they will mess up with some components.

**Reply by leamandeep** [reply]
Ok i will try it

**Reply by leamandeep** [reply]
> Hey i have tried but it isn't working . Please elaborate it

**Reply by kwiat1990** [reply]
What's not working exactly? I have myself have started with that combo and the basics were working. What I haven't yet tried is incorporating Tailwind's config in the theme provider so that mantime use Tailwind's color palette and other stuff like spacing or shadows.

**Reply by ilhamgepe** [reply]
hello sir @rtivital.  can you help me. im using nextjs, if i dont put tailwind base in global.css. here they come
![image](https://user-images.githubusercontent.com/67534909/189475232-8d77dd93-feb2-4282-a61b-058cb32e6e0d.png)

i mark with yellow color. btw its safe if i use css reset here[https://meyerweb.com/eric/tools/css/reset/](url)

if i use css reset its work fine. is this safe for the future?

**Reply by rtivital** [reply]
@ilhamgepe https://mantine.dev/theming/mantine-provider/#css-reset-and-global-styles

**Reply by ilhamgepe** [reply]
> @ilhamgepe https://mantine.dev/theming/mantine-provider/#css-reset-and-global-styles
` <MantineProvider
        withCSSVariables
        withGlobalStyles
        withNormalizeCSS
        theme={{
          colorScheme: "dark",
        }}
      >`

    
    yes sir, I've used that. but it must be helped with css reset by meyerweb. I don't know where I went wrong 🤣

**Reply by PauliCZ44** [reply]
I will just hook into the answer. You can look here:
https://github.com/PauliCZ44/MantineTW

The trick is to use @layer rule.

```
@layer tailwind {
  @tailwind base;
}
@tailwind components;
@tailwind utilities;
```

**Reply by Navin009** [reply]
Worked for me tailwind and mantine button transparent issue is fixed

**Reply by SweydManaf** [reply]
Finally ❤️

**Comment by leamandeep** [comment]
Here's the screenshot of my react project ( just testing )

>App.js file
![image](https://user-images.githubusercontent.com/68345524/176588872-082e5592-8475-499a-b87f-8a98d14d188f.png)

> I haven't used @taiwind base etc stuffs
![image](https://user-images.githubusercontent.com/68345524/176589044-f291eefc-afc9-469a-a111-08eca82eeaee.png)

> tailwind config file
![image](https://user-images.githubusercontent.com/68345524/176589129-f5962c29-3b2d-4be9-9bdd-65bc92db6e14.png)

Mantine is working but tailwind isn't 😞 .

**Reply by kwiat1990** [reply]
Is it CRA project? I see following things without testing it, which you should check: 
* `index.css` doesn’t import any Tailwind layers.
* `index.css` is imported (is it `main.js`?).
* Tailwinds config need  to know about `jsx` because you’ll be using it in those filed and not `html` or `js` files.

**Reply by leamandeep** [reply]
Here's the zip file of my program . I can;t figure out why it is not working 😔

[mantine-tailwind.zip](https://github.com/mantinedev/mantine/files/9018403/mantine-tailwind.zip)

**Reply by kwiat1990** [reply]
@amandeep2603 could you please provide a codesandbox for this? It would be easier to look at the code and how it works.

**Reply by leamandeep** [reply]
Ok

**Reply by kwiat1990** [reply]
@amandeep2603 you need to follow the installation step from here: https://tailwindcss.com/docs/guides/create-react-app. I have run your code and it doesn't work initially but I've noticed that `postcss` and `autoprefixer` were missing as well as `postcss.config.js`. These are required and after I've installed them, Tailwind is up and running.

**Reply by leamandeep** [reply]
Ok thanks i will try it definitely

**Reply by leamandeep** [reply]
here's the link of codesandbox 

https://codesandbox.io/s/mantine-tailwind-w1epus?file=/src/App.js

**Reply by leamandeep** [reply]
> @amandeep2603 you need to follow the installation step from here: https://tailwindcss.com/docs/guides/create-react-app. I have run your code and it doesn't work initially but I've noticed that `postcss` and `autoprefixer` were missing as well as `postcss.config.js`. These are required and after I've installed them, Tailwind is up and running.

Did but still it's not working @kwiat1990

**Reply by kwiat1990** [reply]
@amandeep2603 here's a working demo with your code and missing dependencies: https://github.com/kwiat1990/mantine-tailwind-demo. It works locally. It seems that on Codesanbox Tailwind doesn't run properly.

**Reply by leamandeep** [reply]
Thanks for you precious time @kwiat1990

**Reply by leamandeep** [reply]
Tried it's working 🥳 . Again Thank you 🙏

>But i want to ask what was the reason that it was not working earlier even after i followed you ??

**Reply by kwiat1990** [reply]
@amandeep2603 all I did was to include `postcss` and `autoprefixer` dependencies and then to created `postcss.config.js`. Just like in Tailwind docs. You can compare your code from before and after to see all differences.

**Reply by asher-lab** [reply]
Will follow this procedure as well thanks @kwiat1990

**Reply by omerson-cruz** [reply]
Thanks for the tip @kwiat1990. I also need this :)

**Comment by leamandeep** [comment]
Here's example by @kwiat1990 
>How to Configure tailwind with mantine ?

https://github.com/kwiat1990/mantine-tailwind-demo

**Reply by eigueso** [reply]
Thanks for this example, lets not forget though to leave out the `@tailwind base` from your global css as it will override some Mantine defaults i.e: Buttons will be invisible unless you pass a custom tailwind class.

**Comment by 0xox** [comment]
None of the above solutions worked for us. We had to use a custom Emotion cache with `prepend: false` so that Mantine doesn't load the css at the top of the head.

This works but causes a flash of css on production. For this you need to make sure you are rendering that same emotion Cache during SSR build.

**Reply by Darathor17** [reply]
Hey @0xox ! Thx for the tip! Care to give some hint on how to fix the flash of css? what do you mean make sure we rendering the same emotion cache? (using nextJS 13 if it does matter)

**Comment by AliDayi97** [comment]
I am using Mantine with Tailwind and had no issues until today. I was working on a functionality to scale an image on hover and since using Next Image I used tailwind to configure the hover behaviour via className. The issue I encountered is the scale functionality (transform function to be more specific) didn't work. After a bit of research I found out that '@tailwind base' should be imported in global styles to make it work. The problem is base styles of tailwind overrides/clashes with the Mantine's so I couldn't find a proper solution to use tailwind fully with Mantine given this constraint.

**Reply by rtivital** [reply]
See https://mantine.dev/theming/emotion-cache/

**Reply by AliDayi97** [reply]
I am already using emotion-cache, is there a specific configuration that I am missing to resolve the problem I described above?

**Reply by rtivital** [reply]
https://mantine.dev/theming/emotion-cache/#change-styles-insertion-point

**Reply by AliDayi97** [reply]
I already tried "prepend: false" option, still importing tailwind base would override some of the Mantine styles (for example carousel arrows)

**Reply by rtivital** [reply]
Cannot tell anything without reproduction

**Comment by MiljkovicMarko** [comment]
@rtivital @AliDayi97 I've found, after long hours of trying, maybe even days, that https://mantine.dev/theming/emotion-cache/#server-side-rendering-with-custom-cache doesn't work for me (and possibly others like AliDayi97).
I've found a solution thanks to this issue and fix by same guy [#21](https://github.com/mantinedev/mantine/issues/2119#issuecomment-1229399751)
As he mentions the most important thing is to run the **createEmotionCache** function inside the **_app.tsx** (or in **_document.tsx**?)(which we aren't doing if we follow the mentioned matine docs ssr guide). But unfortunately he didn't make that very clear and buried that by changing other things like insertionPoint, so lost some more time to figure that out...
This works with tailwind base (preflight).
You should also maybe add **important: #app** or something similar to tailwind config (then place that selector in html...), that way you can have tailwind on top of mantine on top of tailwind preflight (tailwind normalize/reset styles)).
**Tested on Mantine v5.10 and Next.js v13**
I've shorten the version of the fix:
```
// emotionCache.ts (in my case; equivalent to rtl-cache.ts in the docs)

import { createEmotionCache, EmotionCache } from '@mantine/core';

let cache: EmotionCache | undefined;

export const emotionCache = () => {
  if (!cache) {
    cache = createEmotionCache({
      key: 'mantine',
      prepend: false,
    });
  }
  return cache;
}
````
use:
````
// _document.tsx
import { emotionCache } from '/src/emotionCache';
import { createStylesServer, ServerStyles } from '@mantine/next';
import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
} from 'next/document';
// ...
const stylesServer = createStylesServer(emotionCache());

export default class _Document extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const initialProps = await Document.getInitialProps(ctx);

    return {
      ...initialProps,
      styles: (
        <>
          {initialProps.styles}
          <ServerStyles
            html={initialProps.html}
            server={stylesServer}
            key="styles"
          />
        </>
      ),
    };
  }

  render() {
 // ...
````

**Reply by cosmtrek** [reply]
This solution works for me. The Mantine docs for ssr guide should be updated.

**Comment by ianshk** [comment]
Sorry to bump this but you probably don't need to remove @base but just set preflight to false in the tailwind config like so:

`export default {
  content: ['./index.html', './src/**/*.{js,ts,jsx,tsx}'],
  corePlugins: {
    preflight: false,
  },
  theme: {
    extend: {},
  },
  plugins: [],
};`

Ref: https://tailwindcss.com/docs/preflight

**Reply by ruslanguns** [reply]
OMG thank you so much!

**Reply by najibghadri** [reply]
This is the right answer

**Reply by hunterwilhelm** [reply]
This worked for me too!

**Reply by gengjiawen** [reply]
`corePlugins: { preflight: false, }` fixed for me, thx.

**Reply by JeremyKennedy** [reply]
Anyone who wants preflight, check this answer: https://github.com/orgs/mantinedev/discussions/1672?sort=top#discussioncomment-6428444

**Comment by ianshk** [comment]
I spent the time last night to make a repro with the latest version of Mantine and Tailwind together. There is a button and slider example in there.

https://github.com/ianshk/MantineTW

**Reply by PauliCZ44** [reply]
I have forked your repo and made some changes that leverages `@layer` rule. That way, one can use preflight with mantine. I did not tested more components than are in the example but maybe this can be a way to go if one wants to use preflight.

https://github.com/PauliCZ44/MantineTW

**Reply by tan-ahmed** [reply]
great work @PauliCZ44  maybe you could create and maintain one universal example which should be on the mantine docs

**Reply by ilhamgepe** [reply]
thanks @ianshk @PauliCZ44 this is my final choose to use tw and mantine.

**Reply by Milkywayrules** [reply]
dang it, this is my chosen approach. Thank you so much @ianshk @PauliCZ44 , and also thank you to others that discuss & givin some alternatives. 🎉

**Reply by mengxi-ream** [reply]
@PauliCZ44 Thank you for your code, you saved me life! 
Now we can use both preflight and mantine, which is important because we don't wannt mess up some component written by tailwindcss (e.g. I don't want my Link has an underline)
According to your solution, we can write our `globals.css` this way:
```css
/* @tailwind base; */
@tailwind components;
@tailwind utilities;

@layer tailwind {
  @tailwind base;
}

/* @import "@mantine/core/styles.layer.css"; */
@import "@mantine/core/styles.css";
@import "@mantine/dates/styles.css";
```

This works magically! But I don't really understand the reason. I've tried another way, which failed, the code is:
```css
/* @tailwind base; */
@tailwind components;
@tailwind utilities;

/* I want to use this line to set tailwind layer defined below to have a lower priority than Mantine, but it doesn't seem to work. don't know why*/
@layer tailwind, mantine

@layer tailwind {
  @tailwind base;
}

@import "@mantine/core/styles.layer.css";
/* @import "@mantine/core/styles.css"; */
@import "@mantine/dates/styles.css";
```
Can someone explain the underlying reason for this? Why does the first piece of code work but the second doesn't?

**Reply by miketdonahue** [reply]
@Crayon-ShinChan - Any luck? I'm also experiencing this solution not working when trying to use the `layer.css` variants from the Mantine CSS.

**Reply by miketdonahue** [reply]
@Crayon-ShinChan I got it. In case you are interested:

```
/* Add @layer to Tailwind Base to prevent Preflight conflicts */
@layer {
  @tailwind base;
}

/* Mantine core */
@import "@mantine/core/styles.css";

/* Mantine foundational components */
@import "@mantine/core/styles/ScrollArea.css";
@import "@mantine/core/styles/UnstyledButton.css";
@import "@mantine/core/styles/VisuallyHidden.css";
@import "@mantine/core/styles/Paper.css";
@import "@mantine/core/styles/Popover.css";
@import "@mantine/core/styles/CloseButton.css";
@import "@mantine/core/styles/Group.css";
@import "@mantine/core/styles/Loader.css";
@import "@mantine/core/styles/Overlay.css";
@import "@mantine/core/styles/ModalBase.css";
@import "@mantine/core/styles/Input.css";
@import "@mantine/core/styles/Flex.css";

/* Mantine components */
@import "@mantine/core/styles/Button.css";

/* Tailwind components & utilities */
@tailwind components;
@tailwind utilities;
```

Assuming the main goal is to allow Tailwind classes to override Mantine styles by doing something like this:

`<Button className="bg-green-600 text-white>Button</Button>`

This solution will work.

**Reply by miketdonahue** [reply]
Better yet, you can also now use the `layers.css` variant with this appropriate. Change the above to:

`@import "@mantine/core/styles.layer.css";`

This provides even greater flexibility.

**Reply by mengxi-ream** [reply]
> Better yet, you can also now use the `layers.css` variant with this appropriate. Change the above to:
> 
> 
> 
> `@import "@mantine/core/styles.layer.css";`
> 
> 
> 
> This provides even greater flexibility.

Layer css doesn't work for me. Only non-layer css does (like your code)

**Reply by TheKnightCoder** [reply]
Here is how I setup my layers

global.css
```
@layer base, mantine, tailwind;

@layer base {
  @tailwind base;
}

@layer tailwind {
  @tailwind components;
  @tailwind utilities;
}
```

layout.tsx file (or file where importing css
```
import './globals.css'
import '@mantine/core/styles.layer.css'

...
```

So the line `@layer base, mantine, tailwind;` defines the order of precedence, and anything outside a layer has precedence over styles within a layer. We want the tailwind base first as this resets the page (setting margins to 0 ect), then mantine which will give us our mantine component styles, and then tailwind components and utilities which allow us to apply our tailwind classes ontop of mantine components.

Also be sure to `layer.css` imports such as `import '@mantine/core/styles.layer.css'` as this wraps the mantine styles into a `mantine` layer. Where the order of precedence for mantine and tailwind is defined in the global.css

**Reply by TheKnightCoder** [reply]
Also don't forget the semicolon when defining the layers `@layer base, mantine, tailwind;`  This may be why @Crayon-ShinChan snippet is not working above

**Reply by Jdruwe** [reply]
@TheKnightCoder thanks for the awesome explanation, this works well. I learned a lot about CSS layers now. I wonder why your answer has no upvotes, let me change that ;)

**Reply by Jdruwe** [reply]
@TheKnightCoder in my case, ring classes are now broken, are they still working for you? Even if I remove Mantine it's still an issue.

**Comment by TruePadawan** [comment]
`@layer {
  @tailwind base;
}`
This fixed it for me

**Reply by specialOne-coder** [reply]
Thank you for that, you saved my day. Disabling preflight or removing {@tailwind base} destroyed some styles for me.

**Reply by enchorb** [reply]
fixed it for me as well on mantine v7. anyone know why this works?

**Reply by PauliCZ44** [reply]
Here is an exaplanation

https://github.com/PauliCZ44/MantineTW/blob/main/src/index.css

**Reply by ceciiiron** [reply]
I am using mantine v7 and `@layer { @tailwind base; }` worked for me. We need further explanation because according to [tailwind documentation](https://tailwindcss.com/docs/functions-and-directives#layer), valid layers are only base, components, and utilities.  @PauliCZ44 added `@layer tailwind {}` which is not on the list, and it still works.

**Reply by PauliCZ44** [reply]
This custom layer has nothing to do with tailwind specifically. It is a CSS feature. You could name as you want and it would work. Real reason that this works is that if you wrap tailwind base into the layer, it will have "lower specifcitiy"(*) than other styles. So mantine styles "wins" but tailwind base still can be applied where mantine styles are not present.

(*) Actually layers do not modify specificity, in real its just something like "order of precedence". 

MDN: https://developer.mozilla.org/en-US/docs/Web/CSS/@layer

Support is now at ~89%~ 92.4% global (according to caniuse).

**Reply by kimhwanhoon** [reply]
I only knew preflight false but thanks to you I can still keep the preflight many thanks!

**Reply by ahulterstrom** [reply]
You are a legend! This information really should be made more available!

**Reply by JeremyKennedy** [reply]
Thank you so much! I really wanted the preflight to work cuz otherwise I'd have to redo a bunch of layout. This worked better than many other solutions!

**Reply by isaacdarcilla** [reply]
This works for me!

**Reply by duyrk** [reply]
this works like a charm thank you

**Reply by jhirn** [reply]
When using this I get this warning

```@import must precede all other statements (besides @charset or empty @layer)```

I think it's just a warning but not sure if it will cause problems. Too soon to tell at this point.

**Reply by nuruddinshohan83** [reply]
@PauliCZ44, your forked setup works great! I've tested loads of configurations from this Github Issue, but yours nailed it. Thanks a bunch!

**Reply by parnexcodes** [reply]
It doesn't work for me, I'm using Page router in NextJS + TailwindCSS.

**Comment by BenSilver1337** [comment]
Thanks, that helped me too!

**Comment by vKongv** [comment]
Hi, i would like to style with tailwind for non-mantine components, is there a way for us to sync the variables used in mantine (eg: spacing size, colors, etc) to tailwind so both have consistent style?

**Reply by Sajarin-M** [reply]
Consider creating a custom Tailwind plugin to for this.
I've developed one and made it available on npm as [tailwind-plugin-mantine](https://www.npmjs.com/package/@devoss/tailwind-plugin-mantine). Please note that it's currently experimental, and its effectiveness may vary. 
Feel free to check it or use the [source code](https://github.com/Sajarin-M/tailwind-plugin-mantine/blob/main/src/index.ts) to create your own solution.
Your feedback would be appreciated for improvements.

**Reply by zmrfzn** [reply]
@Sajarin-M are there any known issues that you can highlight? 

also do you have a readme or example?

**Reply by Sajarin-M** [reply]
Thanks for your interest!
So far, I haven't come across any issues with the plugin. I haven't created one yet as I primarily developed this plugin for personal use.
However, your feedback is valuable, and if there's any interest and feedback, I'll certainly create a readme and example to provide more guidance on usage.

**Comment by TheKnightCoder** [comment]
Sorry I couldn't say as I'm no longer using this in my projects, maybe
someone else can help with this?

**Comment by songkeys** [comment]
edit: Tailwind v4 is now supported! It'll be really easy to integrate:

```css
/** with just one line import */
@import "tailwind-preset-mantine";
```

check out https://github.com/Songkeys/tailwind-preset-mantine

---

I've created a package [`tailwind-preset-mantine`](https://github.com/Songkeys/tailwind-preset-mantine) to seamlessly integrate Mantine styles with Tailwind:

```ts
// `bg-red-500` will be `background-color: var(--mantine-color-red-5)`
// `text-white` will be `color: var(--mantine-color-white)`
<Button className="bg-red-500 text-white">Hello</Button>
```

Additionally, I've put together a guide on how to use `@layer` to resolve style conflicts. A quick overview:

```css
@layer tw_base, mantine, tw_components, tw_utilities;

/* import tailwind */
@import "tailwindcss/base" layer(tw_base);
@import "tailwindcss/components" layer(tw_components);
@import "tailwindcss/utilities" layer(tw_utilities);

/* import mantine */
@import "@mantine/core/styles.layer.css";
```

For more details, check out https://github.com/Songkeys/tailwind-preset-mantine#prevent-style-conflicts.

I've been using this layering method in many of my projects for months, and it works wonderfully!

A minimal template: https://github.com/songkeys/next-app-mantine-tailwind-template

**Reply by Jdruwe** [reply]
@songkeys that's awesome :) I did something similar on my own but only for colors, yours is more complete. Since you mentioned using the layering method in multiple projects I also assume you have been using your library in some way? 

I did notice https://github.com/songkeys/tailwind-preset-mantine/blob/main/src/index.js using `darkMode: ["class", '[data-mantine-color-scheme="dark"]'],` I think selector replaced class: https://tailwindcss.com/docs/dark-mode

**Reply by songkeys** [reply]
@Jdruwe 

> I also assume you have been using your library in some way?

I use this preset in many internal products within my team. I just extracted it into this library today.

And thank you for pointing the darkMode selector issue. I've just [fixed](https://github.com/songkeys/tailwind-preset-mantine/commit/4777ffdd61ebfe6658181d0f4b8addd8d6748580) it. :)

**Reply by Jdruwe** [reply]
> @Jdruwe
> 
> > I also assume you have been using your library in some way?
> 
> I use this preset in many internal products within my team. I just extracted it into this library today.
> 
> And thank you for pointing the darkMode selector issue. I've just [fixed](https://github.com/songkeys/tailwind-preset-mantine/commit/4777ffdd61ebfe6658181d0f4b8addd8d6748580) it. :)

Damn you are fast, awesome work ;)

**Reply by Jdruwe** [reply]
@songkeys I have no clue why but I tried your library but it's not working seems like the classes are not going anything. Probably something on my end. I think it might be related to https://github.com/Songkeys/tailwind-preset-mantine#prevent-style-conflicts

Update: seems like I was missing postcss-import, I did see it in your config but it's required to make the layer(... syntax work, maybe also highlight in green?

**Reply by songkeys** [reply]
@Jdruwe

I think `postcss-import` is already included in `tailwindcss`?

<img width="420" alt="SCR-20240929-milp" src="https://github.com/user-attachments/assets/a2e96f8e-6dcf-4d51-874d-acd176de24e3">

Here’s a minimal template: <https://github.com/songkeys/next-app-mantine-tailwind-template>. I can get it to work without needing to install it. Please let me know if there’s anything I might have overlooked.

Feel free to leave any issues in that repository if you have any questions. I’d like to avoid disturbing those who are subscribed to this discussion.

**Reply by pc-gs** [reply]
> @Jdruwe
> 
> I think `postcss-import` is already included in `tailwindcss`?
> 
> <img alt="SCR-20240929-milp" width="420" src="https://private-user-images.githubusercontent.com/22665058/371826534-a2e96f8e-6dcf-4d51-874d-acd176de24e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwMzE4OTgsIm5iZiI6MTczMDAzMTU5OCwicGF0aCI6Ii8yMjY2NTA1OC8zNzE4MjY1MzQtYTJlOTZmOGUtNmRjZi00ZDUxLTg3NGQtYWNkMTc2ZGUyNGUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI3VDEyMTk1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxMDk4OTMwNzUzNzc0OWZkYWZiNTdjZmRlODFmYzNhOTkzMzM1OTk4NTQ0M2UwNTIxMmNjOWE0YmEwYmY0M2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LZ1_Fm7GBp_0Iny2j5evVk4umJ3923F7hqXSvFMSh80">
> Here’s a minimal template: https://github.com/songkeys/next-app-mantine-tailwind-template. I can get it to work without needing to install it. Please let me know if there’s anything I might have overlooked.
> 
> Feel free to leave any issues in that repository if you have any questions. I’d like to avoid disturbing those who are subscribed to this discussion.

Thank you @songkeys , it's working like a charm ❤️ 

I was trying few days ago a solution to combine the Mantine specs with Tailwind specs, and here is a great solution for that.

<img width="1264" alt="image" src="https://github.com/user-attachments/assets/695b76fa-ef19-492d-89c8-b537e2437be9">

**Reply by songkeys** [reply]
@artemchs Hi. It'd be helpful if you could provide the whole project setup. Could you please use this [minimal template](https://github.com/songkeys/next-app-mantine-tailwind-template) to reproduce this error and raise the issue at https://github.com/songkeys/tailwind-preset-mantine/issues ? Let's make this thread clean. Thank you so much.

edit: it looks like you are using the tailwind `themeConfig` variable in your mantine `theme`? Please check the minimal template mentioned above. You should use mantine's `theme` created by the `createTheme` method.

**Reply by ghost** [reply]
@songkeys I don't know why, but I also had problems with the postcss layer(abc) syntax. They were compiled to "@media" rules. 

```nix
yarn why postcss-import
└─ tailwindcss@npm:3.4.16
   └─ postcss-import@npm:15.1.0 [ed67e] (via npm:^15.1.0 [ed67e])
```

**Comment by Igor-Arefev** [comment]
I'm using Mantine 7.13.4, tailwind ^3.4.14 and next js 14.2.11. 

I am facing a problem that the Slider component on Safari is running very slow in both dev and prod. But works good on Chrome. Has anyone encountered this issue? 

global.css 

```
@import '@mantine/core/styles.css';
@import '@mantine/dates/styles.css';
@import '@mantine/carousel/styles.css';

@layer tailwind {
  @tailwind base;
}

@tailwind components;
@tailwind utilities;`
```

tailwind.config.js 

```
import type { Config } from 'tailwindcss';

import { mantine_tw_preset } from './tailwind.mantine.preset';


const config: Config = {
  darkMode: ['class', '[data-mantine-color-scheme="dark"]'],
  presets: [mantine_tw_preset],
  content: [
    './src/app/**/*.{js,ts,jsx,tsx,mdx}',
    './src/entities/**/*.{js,ts,jsx,tsx,mdx}',
    './src/features/**/*.{js,ts,jsx,tsx,mdx}',
    './src/widgets/**/*.{js,ts,jsx,tsx,mdx}',
    './src/shared/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  plugins: [],
};
export default config;
```

postcss.config.js 

```
module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
    'postcss-preset-mantine': {},
    'postcss-simple-vars': {
      variables: {
        'mantine-breakpoint-xs': '36em',
        'mantine-breakpoint-sm': '48em',
        'mantine-breakpoint-md': '62em',
        'mantine-breakpoint-lg': '75em',
        'mantine-breakpoint-xl': '88em',
      },
    },
  },
};

```

**Comment by Visualizeit** [comment]
TailwindCSS V4 + Mantine V7 Example:

**app.css**

```css
@layer theme, base, mantine, components, utilities;

@import 'tailwindcss';

@import '@mantine/core/styles.layer.css';
```

**Comment by realfresh** [comment]
For anyone struggling to make the above solutions work in a way which:
1. allows your overrides provided to MantineProvider to work
2. allow you to override component styles with tailwind utility classes

Try this out. I haven't tested it fully but it seems to work so far:

Your CSS file:
```css
/**
 *
 * ~----------------------------------------~
 * ~~ Using TailwindCSS with Mantine
 * ~----------------------------------------~
 *
 * ## Requirements:
 * - Use the `@layer` directive to define the order of the layers
 * - Use Mantines `.layer.css` files which define styles in the `mantine` layer
 *
 * ## In order of priority, the layers are:
 * 1. layer(_root_): Tailwind `components` and `utilities`
 * 2. layer(mantine): Mantine's styles
 * 3. layer(tw-base): Tailwind's `base` styles
 *
 * ## Doing this allows:
 * - Tailwind base/preflight styles to be included without conflicting with mantine
 * - Tailwind utility classes to override mantine component styles without using `!important`
 *
**/

/* 1. Define the layer order */
@layer tw-base, mantine;

/* 2. Import Mantine styles defined in `mantine` layer */
@import "@mantine/core/styles.layer.css";
@import "@mantine/dates/styles.layer.css";
@import "@mantine/dropzone/styles.layer.css";
@import "@mantine/notifications/styles.layer.css";

/* 3. Import Tailwind preflight styles into lowest priority layer */
@layer tw-base {
  @tailwind base;
}

/* 4. Import Tailwind components and utilities into root layer (highest priority) */
@tailwind components;
@tailwind utilities;
```

Then on your `MantineProvider`, your `cssVariableSelector` should be:
```tsx
const providerProps = {
  /**
   * To use Mantine with TailwindCSS, Tailwind's base styles need to be on layer
   *  below Mantine's styles (`@layer tw-base, mantine`).
   *
   * However, by doing this, the variables defined in Mantine's stylesheets
   *  `@mantine/core/styles.layer.css` get put under `:root:not(#\#)`.
   *
   * And since the values defined in the `MantineProvider` theme get put into
   *  CSS vars on `:root`, the result is they get overridden by the
   *  defaults from Mantine's stylesheet.
   *
   * To fix this, we must set `cssVariablesSelector` (default is `:root`) to have
   *  higher specificity than the `mantine` layer.
   */
  cssVariablesSelector: `:root${":not(#\\#)".repeat(3)}`,
}
```

**Comment by ghost** [comment]
For all those who still have problems with the layers (in combination with Mantine, Tailwind CSS + Next.js) and have not found a working guide.
My problem was that the layer order didn't work at all. The reason for this was, that the compiler was not quite up to speed at some point and always added the tailwind layer imports to the end of the generated file. I couldn't find out exactly where the error was or which package was responsible for it. However, I was able to work around this bug by extracting the tailwind layers into a new file. 

> global.css

```css
/* 1. Define the layer order */
@layer tw-base, mantine, mantine-datatable, tw-components, tw-utilities;

/* 2. Create new layers for tailwindcss */
@import './tailwind';

/* 3. Import existing layers */
@import '@mantine/core/styles.layer.css'; /* mantine */
@import '@mantine/tiptap/styles.layer.css'; /* mantine */
@import 'mantine-datatable/styles.layer.css'; /* mantine-datatable */
```

> tailwind.css

```css
@layer tw-base {
  @tailwind base;
}

@layer tw-components {
  @tailwind components;
}

@layer tw-utilities {
  @tailwind utilities;
}
```

**Comment by dan-scrumsdotcom** [comment]
Anybody had any luck with Tailwind v4 + Mantine?

**Reply by songkeys** [reply]
```css
/** with just one line import */
@import "tailwind-preset-mantine";
```

check out https://github.com/Songkeys/tailwind-preset-mantine

**Comment by muco-rolle** [comment]
Hey @dan-scrumsdotcom Yes, here's how I'm importing the CSS files. 

```css
// main.css

@layer theme, base, components, mantine, utilities;

@import 'tailwindcss';


@layer mantine {
  @import '@mantine/core/styles.layer.css';
}

```

**Comment by dan-scrumsdotcom** [comment]
All good thanks guys - won't be using Mantine unfortunately, the team were not nice on their community Discord and I won't be supporting that

**Comment by m-torin** [comment]
People have unrealistic expectations for community help, an issue in any open-source project. I've repeatedly returned to this thread on different projects, and these solutions can work. 

Thank you all for your advice and support!!!

**Reply by Jdruwe** [reply]
I agree, this is open source, we are relying on the goodwill of others. Speaking from my experiences I've received a lot of help from others.

**Reply by dan-scrumsdotcom** [reply]
The solutions here definitely can work 💪 

Unfortunately the maintainers were straight up rude on Discord, I wasn't even asking for help or anybodies time.

**Comment by hussain7college** [comment]
there is a fastforward working solution, just do these steps:
1- run: `npm install tailwindcss postcss @tailwindcss/postcss --save-dev`
2- use: https://github.com/songkeys/tailwind-preset-mantine

**Reply by Nifled** [reply]
Just coming through to confirm this is what helped me. Spent some hours trying to figure out why some styles weren't working. 
Regarding this, installing `postcss @tailwindcss/postcss` are not necessary, just standalone `tailwindcss` + tailwindcss vite plugin if you're using that

**Comment by vikito755** [comment]
I know this is resolved, but I was curious. Why would people use Mantine along with Tailwind? What are the use cases?
(genuinely curious)

Mantine looks like a complete solution and extensible with "*.module.css" files, where we can use the theme variables.

**Reply by codejanovic** [reply]
I personally just dont have enough mental capacity left to treat styles like its own enterprise language. 

I usually just want to align things into a flex or grid without a hassle and then go back working on the complex parts of my application again.
